<a href="https://colab.research.google.com/github/RavindraGR7/BIG-FIX/blob/main/Lab8_tej.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import sklearn.datasets as sk
import sklearn.neighbors as kn
(X,y) = sk.load_svmlight_file("colon_cancer_train_1.libsvm")
(X_test,y_test) = sk.load_svmlight_file("colon_cancer_test_1.libsvm")
knn_c = kn.KNeighborsClassifier(n_neighbors=3)
knn_c.fit(X,y)
acc = knn_c.score(X_test,y_test)

checking for accuracy


In [2]:
print(acc)

0.782608695652174


Call the score() function over the training instances (X,y). What is the obtained accuracy?

In [3]:
import sklearn.datasets as sk
import sklearn.neighbors as kn

# Load the dataset
(X, y) = sk.load_svmlight_file("colon_cancer_train_1.libsvm")

# KNN Classifier with n_neighbors=3
knn_c = kn.KNeighborsClassifier(n_neighbors=3)

# Fit the model on the training data
knn_c.fit(X, y)

# Calculate accuracy on the training data
training_acc = knn_c.score(X, y)

print("Accuracy on the training data:", training_acc)


Accuracy on the training data: 0.8205128205128205


In [5]:
import sklearn.datasets as sk
import sklearn.neighbors as kn
from sklearn.model_selection import train_test_split

# Load the dataset
(X, y) = sk.load_svmlight_file("colon_cancer_train_1.libsvm")
X_dense = X.toarray()  # Convert to dense

# Split the data for evaluation
X_train, X_test, y_train, y_test = train_test_split(X_dense, y, test_size=0.2, random_state=42)

n_neighbors = [3, 5, 7, 9, 11]
metrics = sklearn.neighbors.VALID_METRICS['brute']
metrics = [metric for metric in metrics if metric != "precomputed"]

# Initialize variables to track the best performance
best_test_accuracy = 0
best_test_params = {'n_neighbors': None, 'metric': None}

best_train_accuracy = 0
best_train_params = {'n_neighbors': None, 'metric': None}

for n in n_neighbors:
    for m in metrics:
        try:
            knn_c = kn.KNeighborsClassifier(n_neighbors=n, metric=m)
            knn_c.fit(X_train, y_train)

            # Check accuracy on the training set
            train_acc = knn_c.score(X_train, y_train)
            if train_acc > best_train_accuracy:
                best_train_accuracy = train_acc
                best_train_params = {'n_neighbors': n, 'metric': m}

            # Check accuracy on the test set
            test_acc = knn_c.score(X_test, y_test)
            if test_acc > best_test_accuracy:
                best_test_accuracy = test_acc
                best_test_params = {'n_neighbors': n, 'metric': m}

        except Exception as e:
            continue  # If a combination fails, continue to the next one

# Output the results
print(f"Best Training Accuracy: {best_train_accuracy}, Parameters: {best_train_params}")
print(f"Best Test Accuracy: {best_test_accuracy}, Parameters: {best_test_params}")


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/pairwise.py:2025: DataConversionWarning: Data was converted to boolean for metric dice
  warnings.warn(msg, DataConversionWarning)
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/pairwise.py:2025: DataConversionWarning: Data was converted to boolean for metric dice
  warnings.warn(msg, DataConversionWarning)
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/pairwise.py:2025: DataConversionWarning: Data was converted to boolean for metric jaccard
  warnings.warn(msg, DataConversionWarning)
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/pairwise.py:2025: DataConversionWarning: Data was converted to boolean for metric jaccard
  warnings.warn(msg, DataConversionWarning)
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/pairwise.py:2025: DataConversionWarning: Data was converted to boolean for metric kulsinski
  warnings.warn(msg, DataConversionWarning)
/usr/local/lib/python3.10/dist-packages/sklearn/metric

Best Training Accuracy: 0.9354838709677419, Parameters: {'n_neighbors': 3, 'metric': 'hamming'}
Best Test Accuracy: 0.875, Parameters: {'n_neighbors': 7, 'metric': 'chebyshev'}


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/pairwise.py:2025: DataConversionWarning: Data was converted to boolean for metric matching
  warnings.warn(msg, DataConversionWarning)
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/pairwise.py:2025: DataConversionWarning: Data was converted to boolean for metric matching
  warnings.warn(msg, DataConversionWarning)
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/pairwise.py:2025: DataConversionWarning: Data was converted to boolean for metric rogerstanimoto
  warnings.warn(msg, DataConversionWarning)
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/pairwise.py:2025: DataConversionWarning: Data was converted to boolean for metric rogerstanimoto
  warnings.warn(msg, DataConversionWarning)
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/pairwise.py:2025: DataConversionWarning: Data was converted to boolean for metric russellrao
  warnings.warn(msg, DataConversionWarning)
/usr/local/lib/python3.10/dist-

Best Training Accuracy: 0.9354838709677419, Parameters: {'n_neighbors': 3, 'metric': 'hamming'}
Best Test Accuracy: 0.875, Parameters: {'n_neighbors': 7, 'metric': 'chebyshev'}

1)What is the maximum obtained accuracy on the test set? What are the parameter values
(number of neighbors and distance metric) associated with this accuracy?


The maximum obtained accuracy on the test set is 87.5%.
The parameters associated with this accuracy are 7 neighbors and the 'chebyshev' distance metric.

2)What is the maximum obtained accuracy on the training set? What are the parameter
values associated with this accuracy?

The maximum obtained accuracy on the training set is 93.54838709677419%.
The parameters associated with this accuracy are 3 neighbors and the 'hamming' distance metric.

3)Are the parameter values achieving maximum accuracy the same for the training and test
sets?

No, the parameter values achieving maximum accuracy are not the same for the training and test sets. For the training set, the best parameters were 3 neighbors with the 'hamming' metric, while for the test set, the best parameters were 7 neighbors with the 'chebyshev' metric.

4)If no, what does this tell you about the classification performance of the KNN classifier?

The difference in parameter values leading to maximum accuracy on the training vs. test sets suggests that the optimal configuration of the KNN classifier can vary depending on the dataset (or subset of the dataset) being considered. Specifically:
The 'hamming' distance metric with fewer neighbors (3) was more effective for the training set, possibly because it fits the training data more closely.
For the test set, a larger number of neighbors (7) and the 'chebyshev' metric yielded better accuracy, which might indicate a more generalized model that performs better on unseen data.
This discrepancy can also indicate the model's sensitivity to the choice of parameters and the potential for overfitting with certain configurations. Using a higher number of neighbors and a different metric for the test set suggests that a more generalized approach (less sensitive to the noise in the training data) is beneficial for model performance on unseen data.
It highlights the importance of parameter tuning and validation techniques (like cross-validation) to find a balance between fitting the training data well and maintaining good performance on unseen data, thereby improving the model's generalization ability.




